source/Nevermore/Transient/DbCommandExtensions.cs

```diff
@@ -128,39 +84,96 @@ public static async Task<object> ExecuteScalarWithRetryAsync(this DbCommand comm
-                        await command.Connection.CloseAsync();
-                }
+                await using var validCommand = await ValidCommandAsync.For(command, connectionRetryPolicy, cancellationToken);
+                return await validCommand.ExecuteScalarAsync(cancellationToken);
             });
         }
 
         static void GuardConnectionIsNotNull(DbCommand command)
         {
-            if (command.Connection == null)
+            if (command?.Connection == null)
                 throw new InvalidOperationException("Connection property has not been initialized.");
         }
 
+        class ValidCommand : IDisposable
+        {
+            DbConnection ourConnection = null;
+            
+            readonly DbCommand innerCommand = null;
+            
+            ValidCommand(DbCommand command, DbConnection connection = null)
+            {
+                innerCommand = command;
+                ourConnection = connection;
+            }
+
+            public static ValidCommand For(DbCommand command, RetryPolicy retryPolicy)
+            {
+                GuardConnectionIsNotNull(command);
+                
+                if (command.Connection.State == ConnectionState.Open) return new ValidCommand(command);
+                
+                command.Connection.OpenWithRetry(retryPolicy);
+                return new ValidCommand(command, command.Connection);
+            }
+
+            public object ExecuteScalar() => innerCommand.ExecuteScalar();
+
+            public DbDataReader ExecuteReader(CommandBehavior behavior) => innerCommand.ExecuteReader(behavior);
+
+            public int ExecuteNonQuery() => innerCommand.ExecuteNonQuery();
+            
+            public void Dispose()
+            {
+                if (ourConnection?.State != ConnectionState.Open) return;
+                ourConnection.Close();
+                ourConnection = null;
+            }
+        }
+
         /// <summary>
         /// Ensures the command either has an existing open connection, or we will open one for it.
         /// </summary>
         /// <returns>True if we opened the connection (indicating we own its lifetime), False if the connection was already open (indicating someone else owns its lifetime)</returns>
-        static bool EnsureValidConnection(DbCommand command, RetryPolicy retryPolicy)
+        class ValidCommandAsync : IAsyncDisposable
         {
-            if (command == null) return false;
+            DbConnection ourConnection = null;
 
-            GuardConnectionIsNotNull(command);
+            readonly DbCommand innerCommand = null;
+
+            ValidCommandAsync(DbCommand command, DbConnection connection = null)
+            {
+                innerCommand = command;
+                ourConnection = connection;
+            }
+            
+            public static async Task<ValidCommandAsync> For(DbCommand command, RetryPolicy retryPolicy, CancellationToken cancellationToken)
```

Comment:
nit: `ForAsync`

Reply:
Reverted this, mainly for the DbDataReader methods which only close in a catch block... I think there is a better pattern for this, but this PR is not the place for it.